### DISCUSSION TITLE
C++ to Python Implementation

### DISCUSSION THREAD

**Discussion by jha-vineet69** [google/or-tools#2368] [Graph and Algorithms] [answered]
How can I use hamiltonian_path and DijkstraShortestPath etc in Python? I have data as a distance matrix. I don't understand the input structure of these functions.

**Comment by lperron** [comment]
Dijkstra is a trivial implementation. You should look up for a better one.
The hamiltonian path is a bit more tricky, but is not wrapped in python.
So, create your own easy to use API in c++, wrap it in python and use that.
Laurent Perron | Operations Research | lperron@google.com | (33) 1 42 68 53
00



Le dim. 31 janv. 2021 à 19:21, Vineet Jha <notifications@github.com> a
écrit :

**Reply by jha-vineet69** [reply]
Can you point me to some resources on how to do this wrapping part?

**Comment by ghost** [comment, marked as answer]
As Laurent said it isn't recommended, but here's an example for Dijkstra anyway:
```python
from ortools.graph import pywrapgraph

graph = [
    [0, 4, 0, 0, 0, 0, 0, 8, 0],
    [4, 0, 8, 0, 0, 0, 0, 11, 0],
    [0, 8, 0, 7, 0, 4, 0, 0, 2],
    [0, 0, 7, 0, 9, 14, 0, 0, 0],
    [0, 0, 0, 9, 0, 10, 0, 0, 0],
    [0, 0, 4, 14, 10, 0, 2, 0, 0],
    [0, 0, 0, 0, 0, 2, 0, 1, 6],
    [8, 11, 0, 0, 0, 0, 1, 0, 7],
    [0, 0, 2, 0, 0, 0, 6, 7, 0],
]

start = 0
end = 8
nodes = len(graph)
dijkstra = pywrapgraph.DijkstraShortestPath(
    nodes, start, end, lambda i, j: graph[i][j], disconnected_distance=0
)
print(dijkstra)

```

**Reply by jha-vineet69** [reply]
Thanks, @stradivari96.
